core/state/snapshot/generate.go

```diff
@@ -159,169 +178,313 @@ func journalProgress(db ethdb.KeyValueWriter, marker []byte, stats *generatorSta
+			default:
+			}
+			if batch.ValueSize() > ethdb.IdealBatchSize || abort != nil {
+				// Flush out the batch anyway no matter it's empty or not.
+				// It's possible that all the states are recovered and the
+				// generation indeed makes progress.
 				marker := accountHash[:]
 				journalProgress(batch, marker, stats)
 
-				batch.Write()
+				if err := batch.Write(); err != nil {
+					return err
+				}
 				batch.Reset()
 
 				dl.lock.Lock()
 				dl.genMarker = marker
 				dl.lock.Unlock()
+
+				if abort != nil {
+					stats.Log("Aborting state snapshot generation", dl.root, accountHash[:])
+					return errors.New("aborted")
+				}
 			}
-			if abort != nil {
-				stats.Log("Aborting state snapshot generation", dl.root, accountHash[:])
-				abort <- stats
-				return
-			}
-		}
-		// If the account is in-progress, continue where we left off (otherwise iterate all)
-		if acc.Root != emptyRoot {
-			storeTrie, err := trie.NewSecure(acc.Root, dl.triedb)
-			if err != nil {
-				log.Error("Generator failed to access storage trie", "root", dl.root, "account", accountHash, "stroot", acc.Root, "err", err)
-				abort := <-dl.genAbort
-				abort <- stats
-				return
-			}
-			var storeMarker []byte
-			if accMarker != nil && bytes.Equal(accountHash[:], accMarker) && len(dl.genMarker) > common.HashLength {
-				storeMarker = dl.genMarker[common.HashLength:]
-			}
-			storeIt := trie.NewIterator(storeTrie.NodeIterator(storeMarker))
-			for storeIt.Next() {
-				rawdb.WriteStorageSnapshot(batch, accountHash, common.BytesToHash(storeIt.Key), storeIt.Value)
-				stats.storage += common.StorageSize(1 + 2*common.HashLength + len(storeIt.Value))
-				stats.slots++
-
-				// If we've exceeded our batch allowance or termination was requested, flush to disk
-				var abort chan *generatorStats
-				select {
-				case abort = <-dl.genAbort:
-				default:
+			// If the iterated account is the contract, create a further loop to
+			// verify or regenerate the contract storage.
+			if acc.Root != emptyRoot {
+				var storeMarker []byte
+				if accMarker != nil && bytes.Equal(accountHash[:], accMarker) && len(dl.genMarker) > common.HashLength {
+					storeMarker = dl.genMarker[common.HashLength:]
 				}
-				if batch.ValueSize() > ethdb.IdealBatchSize || abort != nil {
-					// Only write and set the marker if we actually did something useful
-					if batch.ValueSize() > 0 {
-						// Ensure the generator entry is in sync with the data
-						marker := append(accountHash[:], storeIt.Key...)
-						journalProgress(batch, marker, stats)
-
-						batch.Write()
-						batch.Reset()
-
-						dl.lock.Lock()
-						dl.genMarker = marker
-						dl.lock.Unlock()
+				var storeOrigin = common.CopyBytes(storeMarker)
+				for {
+					exhausted, last, err := dl.genRange(acc.Root, append(rawdb.SnapshotStoragePrefix, accountHash.Bytes()...), "storage", storeOrigin, storageCheckRange, stats, func(key []byte, val []byte, regen bool) error {
+						if regen {
```

Comment:
OMG the indentation level here is getting ridiculous. Can e.g. this function definition be lifted out or made standalone?

Reply:
Yes, it's very huge function, I will try to polish it a bit.